stablehlo/dialect/TypeInference.cpp

```diff
@@ -866,15 +878,30 @@ LogicalResult inferPadOp(
       return emitOptionalError(
           location,
           "Interior padding cannot be negative: ", paddingInteriorVal);
-    int64_t expectedOutput =
-        inputShape[i] + paddingLowVal + paddingHighVal +
-        std::max<int64_t>(inputShape[i] - 1, 0LL) * paddingInteriorVal;
-    if (expectedOutput < 0)
-      return emitOptionalError(
-          location, "Padding result in negative size for dimension ", i);
-    resultShape.push_back(expectedOutput);
+
+    if (!hlo::isDynamicDimSize(inputShape[i])) {  // static dim
+      int64_t expectedOutput = inferPadDim(inputShape[i], paddingLowVal,
+                                           paddingHighVal, paddingInteriorVal);
+      if (expectedOutput < 0)
+        return emitOptionalError(
+            location, "Padding result in negative size for dimension ", i);
+      resultShape.push_back(expectedOutput);
+
+      if (!inputBounds.empty())
+        resultBounds.push_back(ShapedType::kDynamicSize);
+    } else {  // dynamic dim with bounds
+      int64_t resultBound = inferPadDim(inputBounds[i], paddingLowVal,
+                                        paddingHighVal, paddingInteriorVal);
+      if (resultBound < 0 && resultBound != ShapedType::kDynamicSize)
+        return emitOptionalError(
+            location, "Padding result in negative size for bound of dimension ",
+            i);
+      resultBounds.push_back(resultBound);
+    }
```

Comment:
I think it's a great idea to share logic between the cases which compute results for static dimension sizes and for bounds of dynamic dimension sizes. However, I think that the amount of code in this PR to implement this is fairly excessive.

Here's the code that we want to share:

```C++
int64_t expectedOutput =
    inputShape[i] + paddingLowVal + paddingHighVal +
    std::max<int64_t>(inputShape[i] - 1, 0LL) * paddingInteriorVal;
if (expectedOutput < 0)
  return emitOptionalError(
      location, "Padding result in negative size for dimension ", i);
resultShape.push_back(expectedOutput);
```

And here's how it can be rewritten, without needing additional lambdas and ifs:

```C++
// Assuming that `bool isDynamicDim = hlo::isDynamicDimSize(inputShape[i]);` is declared above.
int64_t operandSizeOrBound = isDynamicDim ? inputBounds[i]: inputShape[i];
int64_t resultSizeOrBound =
    operandSizeOrBound + paddingLowVal + paddingHighVal +
    std::max<int64_t>(inputShape[i] - 1, 0LL) * paddingInteriorVal;
if (resultSizeOrBound < 0) {
  auto what = isDynamicDim ? "bound" : "size";
  return emitOptionalError(
      location, "Padding result in negative ", what, " for dimension ", i);
}
(isDynamicDim ? resultBounds : resultShape).push_back(expectedOutput);
```

Reply:
The new commit is a lot simplified by the comment above and the long discussion result from https://github.com/openxla/stablehlo/pull/378.  The keys are:
1. Reuse code for `isStaticDim` case and `isStaticBound` case
2. Init the result_shape and result_bounds with (len, ShapedType::kDynamicSize) which avoid a lot of "appending"s in complicated different cases.